### PR TITLE
Fixes for correctness and short-circuiting of Stream#{take,repeat,repeatElems}

### LINKED ISSUE #892: Rename ZStream#repeatElems => ZStream#spaced
Rename the `ZStream#repeatElems` method `ZStream#spaced`.

### PR DESCRIPTION
**Pull request by iravid**
Also resolves #892

### CODE REVIEW

File: streams/shared/src/main/scala/scalaz/stream/ZStream.scala [hunk 1]
```diff
@@ -408,16 +408,19 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
...
-              if (decision.cont)
-                IO.unit.delay(decision.delay) *> f(s, a).flatMap(loop(_, decision.state, a))
-              else IO.succeed(s)
-            }
+            if (!cont(s)) ZIO.succeed(s)
+            else
+              f(s, a).zip(schedule.update(a, sched)).flatMap {
+                case (s, decision) =>
+                  if (decision.cont && cont(s))
+                    IO.unit.delay(decision.delay) *> loop(s, decision.state, a)
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/scalaz/stream/ZStream.scala]: Can you not just `loop(...).delay(decision.delay)`?

Comment (iravid) [reply streams/shared/src/main/scala/scalaz/stream/ZStream.scala]: Oh yeah. 🤦🏼‍♂️

File: streams/jvm/src/test/scala/scalaz/zio/stream/StreamSpec.scala [hunk 2]
```diff
@@ -572,4 +595,44 @@ class StreamSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
...
+              .run(Sink.drain)
+        result <- ref.get
+      } yield result must_=== List(1, 1)
+    )
+
+  private def spaced =
+    unsafeRun(
+      Stream(1, 2, 3)
+        .spaced(Schedule.recurs(1))
+        .run(Sink.collect[Int])
```
Comment (regiskuckaertz) [inline streams/jvm/src/test/scala/scalaz/zio/stream/StreamSpec.scala]: I'm only starting to see how awesome this API is 😍

Comment (iravid) [reply streams/jvm/src/test/scala/scalaz/zio/stream/StreamSpec.scala]: Going to be even nicer with the upcoming restructuring around `ZManaged` :-)

File: streams/shared/src/main/scala/scalaz/stream/ZStream.scala [hunk 3]
```diff
@@ -480,7 +483,19 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
...
-    self.zipWithIndex.collectWhile { case (v, i) if i < n => v }
+    if (n <= 0) Stream.empty
+    else
+      new ZStream[R, E, A] {
+        override def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] =
+          IO.succeedLazy { (s, cont, f) =>
+            self.zipWithIndex.fold[R1, E1, (A, Int), (S, Boolean)].flatMap { f0 =>
+              f0(s -> true, tp => cont(tp._1) && tp._2, {
+                case ((s, _), (a, i)) =>
+                  f(s, a).map((_, if (i == n - 1) false else true))
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/scalaz/stream/ZStream.scala]: just `i < n-1`?

Comment (iravid) [reply streams/shared/src/main/scala/scalaz/stream/ZStream.scala]: Right! Thank you for the thorough review :-)

**Comment by iravid** [comment]
Thanks for reviewing @regiskuckaertz!

**Pull request merged**